lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -22,90 +22,156 @@
+
+    /// Recursively computes the transitive closure over members
+    /// adding memberOperatorDecls and extendedNominals to the receiver.
+    void findNominalsAndOperators(const DeclRange members);
+  };
 
-static void findNominalsAndOperators(
-    llvm::MapVector<const NominalTypeDecl *, bool> &foundNominals,
-    llvm::SmallVectorImpl<const FuncDecl *> &foundOperators,
-    DeclRange members) {
-  for (const Decl *D : members) {
-    auto *VD = dyn_cast<ValueDecl>(D);
-    if (!VD)
-      continue;
-
-    if (VD->hasAccess() &&
-        VD->getFormalAccess() <= AccessLevel::FilePrivate) {
-      continue;
-    }
-
-    if (VD->getFullName().isOperator()) {
-      foundOperators.push_back(cast<FuncDecl>(VD));
-      continue;
-    }
-
-    auto nominal = dyn_cast<NominalTypeDecl>(D);
-    if (!nominal)
-      continue;
-    foundNominals[nominal] |= true;
-    findNominalsAndOperators(foundNominals, foundOperators,
-                             nominal->getMembers());
-  }
-}
-
-static bool declIsPrivate(const Decl *member) {
-  auto *VD = dyn_cast<ValueDecl>(member);
-  if (!VD) {
-    switch (member->getKind()) {
-    case DeclKind::Import:
-    case DeclKind::PatternBinding:
-    case DeclKind::EnumCase:
-    case DeclKind::TopLevelCode:
-    case DeclKind::IfConfig:
-    case DeclKind::PoundDiagnostic:
-      return true;
-
-    case DeclKind::Extension:
-    case DeclKind::InfixOperator:
-    case DeclKind::PrefixOperator:
-    case DeclKind::PostfixOperator:
-      return false;
-
-    default:
-      llvm_unreachable("everything else is a ValueDecl");
-    }
-  }
-
-  return VD->getFormalAccess() <= AccessLevel::FilePrivate;
-}
-
-static bool extendedTypeIsPrivate(TypeLoc inheritedType) {
-  auto type = inheritedType.getType();
-  if (!type)
-    return true;
-
-  if (!type->isExistentialType()) {
-    // Be conservative. We don't know how to deal with other extended types.
-    return false;
-  }
-
-  auto layout = type->getExistentialLayout();
-  assert(!layout.superclass && "Should not have a subclass existential "
-         "in the inheritance clause of an extension");
-  for (auto protoTy : layout.getProtocols()) {
-    if (!declIsPrivate(protoTy->getDecl()))
-      return false;
-  }
-
-  return true;
-}
+  /// Emit all provided declartions.
+  void emit() const;
+
+  CollectedDeclarations emitTopLevelNames() const;
+  void emitNominalTypes(const llvm::MapVector<const NominalTypeDecl *, bool>
+                            &extendedNominals) const;
+  void emitMembers(const CollectedDeclarations &cpd) const;
+  void emitDynamicLookupMembers() const;
+
+  void emitTopLevelDecl(const Decl *D, CollectedDeclarations &cpd) const;
+  void emitExtensionDecl(const ExtensionDecl *D,
+                         CollectedDeclarations &cpd) const;
+  void emitNominalTypeDecl(const NominalTypeDecl *NTD,
+                           CollectedDeclarations &cpd) const;
+  void emitValueDecl(const ValueDecl *VD) const;
```

Comment:
I'm still unhappy with this because NominalTypeDecls are also ValueDecls, but I can't think of a better answer.